enumerable_methods.rb

```diff
@@ -0,0 +1,163 @@
+# frozen_string_literal: true
+
+module Enumerable # rubocop:disable Metrics/ModuleLength
+  # my_each
+  def my_each
+    return to_enum unless block_given?
+
+    i = 0
+    while i < size
+      yield(self[i])
+      i += 1
+    end
+  end
+
+  # my_each_with_index
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+  end
+
+  # my_select
+  def my_select
+    return to_enum unless block_given?
+
+    arr = []
+    my_each do |a|
+      arr.push(a) if yield(a)
+    end
+    arr
+  end
+
+  # my_all
+  def my_all?(arg = nil) # rubocop:disable Layout/TrailingWhitespace
+    if block_given?
+      my_each { |a| return false unless yield(a) }
+      return true
+    end
+    if arg.class == Regexp || arg.class == String
+      my_each { |a| return false unless a.match(arg) }
+      return true
+    elsif arg.class == Class
+      my_each { |a| return false unless a.class == arg || a.class.superclass == arg }
+      return true
+    elsif arg.nil? && !block_given? && !empty?
+      my_each { |a| return false if a.nil? || a == false }
+    elsif (arg.class != Class || arg.class != Regexp) && !arg.nil?
+      my_each { |a| return true unless a != arg }
+    end
+    true
+  end
```

Comment:
enumerables my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
     Failure/Error: expect(array.my_all?(3)).to be array.all?(3)
     
       expected false
            got true